lib/final-api/model_extensions.rb

```diff
@@ -83,6 +48,68 @@ def sanitize
 
   private
 
+  # Returns list of parsed subqueries
+  #
+  # For example:
+  #   parse_query('nam:"foo bar baz" bui =qux id : 1')
+  #     => [ ['nam', ':', 'foo bar baz'], ['bui', '=', 'qux'], ['id', ':', '1']]
+  def self.parse_query(query)
+    array = query.scan(/([^\s]*)\s*([:=])\s*("[^"]*"|\S*)/)
+    return [['name' , ':', query]] if array.length == 0
+    wrong_keys = []
+    result = array.map do |item|
+      query_key = item[0].downcase
+      column = SEARCH_TOKENS_DEF.select { |k| k.include? query_key }.values.first
+      wrong_keys << query_key if column.nil?
+
+      [
+        column,
+        item[1],
+        item[2].tr("\"", '')
+      ]
+    end
+
+    raise InvalidQueryError,
+          "Wrong search definition(s) specified: #{wrong_keys.join(", ")}" unless wrong_keys.empty?
+
+    result
+  end
+
+  def self.retrieve_users(query, exact_match = false)
+    if exact_match
+      User.where(name: query).each_with_object([]) {|u,out| out << u.id }
+    else
+      User.where("name ILIKE :expr", expr: "%#{query}%").each_with_object([]) {|u,out| out << u.id }
+    end
+  end
+
+  # maps fragment of old state given to travis states
+  def self.determine_states(query, exact_match = false)
+    states_map = FinalAPI::V1::Http::DDTF_Build::BUILD_STATE2API_V1STATUS
+    states_map.reject { |k,v| k == '' }.each_with_object([]) do |(new, old), out|
+      if exact_match
+        out << new if old.downcase == query.downcase
+      else
+        out << new if old.downcase.include? query.downcase
+      end
+    end.compact
+  end
+
+  def self.retrieve_filter(expr)
+    exact_match = (expr[1] == '=')
```

Comment:
expr[0] is used multiple times, which means dereferencing each time, what about:
field = expr[0]
operator = expr[1]
value = expr[2]